core/src/style/textStyleBuilder.cpp

```diff
@@ -209,59 +207,240 @@ bool TextStyleBuilder::addFeature(const Feature& _feat, const DrawRule& _rule) {
+            dir2 = (p1 - p2) / segmentLength;
+
+            glm::vec2 pp = glm::vec2(_line[j-1]);
+            float d = sqPointSegmentDistance(pp, p0, p2);
+            if (d > tolerance) { break; }
+
+            if ((glm::length2(dir1 + dir2) < sqDirLimit) ||
+                (glm::length2(dir0 + dir2) < sqDirLimit)) {
+                break;
+            }
+
+            merged++;
+
+            p1 = p2;
+            dir1 = dir2;
+        }
+
+        // place labels at segment-subdivisions
+        int run = merged > 0 ? 1 : 2;
+        segmentLength /= run;
+
+        while (segmentLength > minLength && run <= 4) {
+            glm::vec2 a = p0;
+            glm::vec2 b = glm::vec2(p1 - p0) / float(run);
+
+            for (int r = 0; r < run; r++) {
+                addLabel(_params, Label::Type::line, {{ a, a+b }}, _rule);
+                a += b;
+            }
+            run *= 2;
+            segmentLength /= 2.0f;
+        }
+
+        if (i == 0 && j == _line.size()) {
+            // Simple straight line
+            return true;
+        }
+
+        // Skip merged segments in outer loop
+        i += merged;
+
+    }
+    return false;
+}
+
+void TextStyleBuilder::addCurvedTextLabels(const Line& _line, const TextStyle::Parameters& _params,
+                                           const DrawRule& _rule) {
+
+    // Size of pixel in tile coordinates
+    const float pixelSize = 1.0/m_tileSize;
+    // length of line needed for the label
+    const float labelLength = m_attributes.width * pixelSize;
+    // Allow labels to appear later than tile's min-zoom
+    const float minLength = labelLength * 0.6;
+
+    // Chord length for minimal ~120 degree inner angles (squared)
+    // sin(60)*2
+    const float sqDirLimit = powf(1.7f, 2);
+    // Range to check for angle changes
+    const float sampleWindow = pixelSize * 50;
+
+    // Minimal ~10 degree counts as change of direction
+    // cross(dir1,dir2) < sin(10)
+    const float flipTolerance = 0.17;
+
+    LineSampler<std::vector<glm::vec3>> sampler;
+
+    sampler.set(_line);
+
+    if (sampler.sumLength() < minLength) { return; }
+
+    struct LineRange {
+        size_t start, end;
+        int flips;
+        float area;
+    };
+
+    std::vector<LineRange> ranges;
+
+    for (size_t i = 0; i < _line.size()-1; i++) {
+
+        int flips = 0;
+        float lastAngle = 0;
+        float sumAngle = 0;
+        size_t lastBreak = 0;
+
+        glm::vec2 dir1 = sampler.segmentDirection(i);
 
-                    float d = sqPointSegmentDistance(pp, p1, p);
-                    if (d > tolerance) { break; }
+        for (size_t j = i + 1; j < _line.size()-1; j++) {
+            glm::vec2 dir2 = sampler.segmentDirection(j);
 
-                    // Skip merged segment
-                    merged = true;
-                    i += 1;
+            float angle = crossProduct(dir1, dir2);
+
+            if (std::abs(angle) > flipTolerance) {
```

Comment:
Yep the rule for 'squiggly' labels is definitely important. The one up there seems to be _almost_ readable, but we can still run into cases where the letters can be upside down. It would be safe to add a rule to ensure that no quads gets upside down in screen space. I will give a look to implement that today.

Cases like this one are quite hard to read:
![screen shot 2017-02-02 at 3 01 19 pm](https://cloud.githubusercontent.com/assets/7061573/22596868/4ecb99b0-e9fb-11e6-83f0-754a6d124486.png)